pkg/ruler/rule_backup_manager.go

```diff
@@ -0,0 +1,150 @@
+package ruler
+
+import (
+	"context"
+	"errors"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/prometheus/model/rulefmt"
+	"github.com/prometheus/prometheus/promql/parser"
+	promRules "github.com/prometheus/prometheus/rules"
+
+	"github.com/cortexproject/cortex/pkg/ruler/rulespb"
+)
+
+// Implements GroupLoader interface but instead of reading from a file when Load is called, it returns the
+// rulefmt.RuleGroup it has stored
+type loader struct {
+	ruleGroups map[string][]rulefmt.RuleGroup
+}
+
+func (r *loader) Load(identifier string) (*rulefmt.RuleGroups, []error) {
+	return &rulefmt.RuleGroups{
+		Groups: r.ruleGroups[identifier],
+	}, nil
+}
+
+func (r *loader) Parse(query string) (parser.Expr, error) {
+	return parser.ParseExpr(query)
+}
+
+// rulesBackupManager is an in-memory store that holds []promRules.Group of multiple users. It only stores the Groups,
+// it doesn't evaluate them.
+type rulesBackupManager struct {
+	inMemoryRuleGroupsBackup map[string][]*promRules.Group
+	cfg                      Config
+
+	logger log.Logger
+
+	backupGroupRulesCount      *prometheus.GaugeVec
+	lastBackupReloadSuccessful *prometheus.GaugeVec
+}
+
+func newRulesBackupManager(cfg Config, logger log.Logger, reg prometheus.Registerer) *rulesBackupManager {
+	return &rulesBackupManager{
+		inMemoryRuleGroupsBackup: make(map[string][]*promRules.Group),
+		cfg:                      cfg,
+		logger:                   logger,
+
+		backupGroupRulesCount: promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: "cortex",
+			Name:      "ruler_backup_rule_group_rules",
+			Help:      "The number of backed up rules",
+		}, []string{"user", "rule_group"}),
+		lastBackupReloadSuccessful: promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: "cortex",
+			Name:      "ruler_backup_last_reload_successful",
+			Help:      "Boolean set to 1 whenever the last configuration reload attempt was successful.",
+		}, []string{"user"}),
+	}
+}
+
+// setRuleGroups updates the map[string][]*promRules.Group that the rulesBackupManager stores in memory.
+func (r *rulesBackupManager) setRuleGroups(_ context.Context, ruleGroups map[string]rulespb.RuleGroupList) {
+	backupRuleGroups := make(map[string][]*promRules.Group)
+	for user, groups := range ruleGroups {
```

Comment:
do we need a mutex lock here? When we update we wait before we finished updating

Reply:
It was pointed out in https://github.com/cortexproject/cortex/pull/5782#discussion_r1532973551 that since syncRules happen sequentially we don't need a lock here.